iothub_client/samples/iothub_client_device_cbor_sample/iothub_client_device_cbor_sample.c

```diff
@@ -0,0 +1,956 @@
+    if (rc != IOTHUB_CLIENT_OK)
+    {
+        (void)printf("Failed to set option for %s: return code %d.\n", OPTION_AUTO_URL_ENCODE_DECODE, rc);
+        exit(rc);
+    }
+
+    // Format ONLY the device twin document using CBOR.
+    // ONLY valid for use with MQTT. Must occur prior to CONNECT.
+    OPTION_TWIN_CONTENT_TYPE_VALUE ct = OPTION_TWIN_CONTENT_TYPE_CBOR;
+    rc = IoTHubDeviceClient_SetOption(iothub_client_handle, OPTION_TWIN_CONTENT_TYPE, &ct);
+    if (rc != IOTHUB_CLIENT_OK)
+    {
+        (void)printf("Failed to set option for %s: return code %d.\n", OPTION_TWIN_CONTENT_TYPE, rc);
+        exit(rc);
+    }
+
+#ifdef SET_TRUSTED_CERT_IN_SAMPLES
+    // Use SDK-supplied trusted certificates to run the sample.
+    // ONLY to be used for the sample. **NOT to be used in production code.**
+    rc = IoTHubDeviceClient_SetOption(iothub_client_handle, OPTION_TRUSTED_CERT, certificates);
+    if (rc != IOTHUB_CLIENT_OK)
+    {
+        (void)printf("Failed to set option for %s: return code %d.\n", OPTION_TRUSTED_CERT, rc);
+        exit(rc);
+    }
+#endif
+}
+
+static void connect_device_client_send_and_receive_messages(void)
+{
+    int rc;
+
+    //
+    // Send and receive messages from IoT Hub. Connection occurs when a message is first sent.
+    //
+
+    // Set callback for connection status to IoT Hub.
+    rc = IoTHubDeviceClient_SetConnectionStatusCallback(iothub_client_handle, connection_status_callback, NULL);
+    if (rc != IOTHUB_CLIENT_OK)
+    {
+        (void)printf("Failed to set the connection status callback: return code %d.\n", rc);
+        exit(rc);
+    }
+
+    // Set callback for GET twin document request. - Approach #1 to requesting the twin document.
+    // Send asynchronous GET request for twin document. Connection will occur here.
+    rc = IoTHubDeviceClient_GetTwinAsync(iothub_client_handle, get_twin_async_callback, NULL);
+    if (rc != IOTHUB_CLIENT_OK)
+    {
+        (void)printf("Failed to send a GET request for the twin document asynchronously: return code %d.\n", rc);
+        exit(rc);
+    }
+
+    // Set callback for twin desired property PATCH messages from IoT Hub. - Approch #2 to requesting the twin document.
+    // Sends GET request for twin document as part of setting the callback.
+    rc = IoTHubDeviceClient_SetDeviceTwinCallback(iothub_client_handle, twin_desired_properties_callback, NULL);
+    if (rc != IOTHUB_CLIENT_OK)
+    {
+        (void)printf("Failed to set the twin document callback: return code %d.\n", rc);
+        exit(rc);
+    }
+
+    // Set callback for C2D messages from the service client application.
+    rc = IoTHubDeviceClient_SetMessageCallback(iothub_client_handle, receive_c2d_message_callback, NULL);
+    if (rc != IOTHUB_CLIENT_OK)
+    {
+        (void)printf("Failed to set the C2D message callback: return code %d.\n", rc);
+        exit(rc);
+    }
+
+    send_telemetry();
```

Comment:
nit: why have here if you call it in the loop below?

Reply:
So that something gets sent before the message timeout or the user has to send a c2d/desire properties patch from the hub.  I could always change the order and have telemetry go first.  just a personal preference.